matchers/compat/parse.go

```diff
@@ -0,0 +1,218 @@
+// Copyright 2023 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package compat
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+
+	"github.com/prometheus/alertmanager/featurecontrol"
+	"github.com/prometheus/alertmanager/matchers/parse"
+	"github.com/prometheus/alertmanager/pkg/labels"
+)
+
+var (
+	parseMatcher  = stableMatcherParser(log.NewNopLogger())
+	parseMatchers = stableMatchersParser(log.NewNopLogger())
+)
+
+type matcherParser func(s string) (*labels.Matcher, error)
+
+type matchersParser func(s string) (labels.Matchers, error)
+
+// Matcher parses the matcher in the input string. It returns an error
+// if the input is invalid or contains two or more matchers.
+func Matcher(s string) (*labels.Matcher, error) {
+	return parseMatcher(s)
+}
+
+// Matchers parses one or more matchers in the input string. It returns
+// an error if the input is invalid.
+func Matchers(s string) (labels.Matchers, error) {
+	return parseMatchers(s)
+}
+
+// InitFromFlags initializes the compat package from the flagger.
+func InitFromFlags(l log.Logger, f featurecontrol.Flagger) {
+	if f.StableMatchersParsing() {
+		parseMatcher = stableMatcherParser(l)
+		parseMatchers = stableMatchersParser(l)
+	} else if f.UTF8MatchersParsing() {
+		parseMatcher = utf8MatcherParser(l)
+		parseMatchers = utf8MatchersParser(l)
+	} else {
+		parseMatcher = fallbackMatcherParser(l)
+		parseMatchers = fallbackMatchersParser(l)
+	}
```

Comment:
```suggestion
	if f.StableMatchersParsing() {
	    parseMatcher = stableMatcherParser(l)
	    parseMatchers = stableMatchersParser(l)
	}
	
        if f.UTF8MatchersParsing() {
	  parseMatcher = utf8MatcherParser(l)
	  parseMatchers = utf8MatchersParser(l)
	}
	
	parseMatcher = fallbackMatcherParser(l)
	parseMatchers = fallbackMatchersParser(l)
```

nit: I find this to be a bit easier to read - there's no way `UTF8MatchersParsing` and `StableMatchersParsing` can both we true.

Reply:
It's not correct though? This code will forever use the fallback parser?